onnx/defs/text/defs.cc

```diff
@@ -33,6 +33,69 @@ ONNX_OPERATOR_SET_SCHEMA(
                 *ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape());
         }));
 
+static const char* StringSplit_doc =
+    R"DOC(StringSplit splits a string tensor's elements into substrings based on a delimiter attribute and a maxsplit attribute. The first output of this operator is a tensor of strings representing the substrings from splitting each input string on the delimiter substring. A integer tensor is also returned representing the number of substrings generated. This output tensor has one additional rank compared to the input to store these substrings, as examples below will illustrate.)DOC";
```

Comment:
May be useful to add the description of padding behavior here too. (I notice that it is described in the Output documentation below.)

Reply:
The last sentence is misleading. `This output tensor` should be changed to `The first output tensor`

Reply:
Tensorflow split documentation says
```
If sep is given, consecutive delimiters are not grouped together and are deemed to delimit empty strings. For example, input of "1<>2<><>3" and sep of "<>" returns ["1", "2", "", "3"]. If sep is None or an empty string, consecutive whitespace are regarded as a single separator, and the result will contain no empty strings at the start or end if the string has leading or trailing whitespace.
```
It would be worth describing the intended behavior for these special cases.

Reply:
That should include empty strings as well, one empty string and another case with only empty strings. Should we specify what happens with null dimensions?

Reply:
On second thought, I think having null/0 trailing dimensions work requires a bit more thought on the shape inference front.

StringSplit builds a new substring tensor for each input. Let's take an example input tensor of shape `(2, 0)`. Here, we have no input elements, so an output substring tensor should have shape `(2, 0)` (the same as the input) - this is also how `np.char.split` works. 

```
>> np.char.split(np.array([[], []], dtype=np.str_))
array([], shape=(2, 0), dtype=object)
```

The issue we run into is that this diverges from the general case for shape inference where we add a dimension to the input shape, getting `(2, 0, None)`, and we reach a point where shape inference is dependent on the value of the final dimension(s) being available - this is pretty undesirable since it means we'd have to block shape inference if these final dimension(s) are symbolic [params](https://github.com/onnx/onnx/blob/5c713e6152d8d46d0cc9d4d3f1db841341668f2d/onnx/onnx.proto3#L665) or unknown. A related fact is that a "pseudo-shape" `(2, 0, X)` is not equivalent to `(2, 0)` with the current tooling (though they are in some respects equal).

All of this is to say, I think we have to restrict shape inference on this output tensor quite significantly to support null dimensions. Any opinions on this option (or alternatively raising)?

(As a side note, I think some Reduce* (e.g. ReduceSum, ReduceL1) operators end up inferring the wrong rank or segfaulting for empty inputs during shape inference as well)

Reply:
Is there any compelling reason to stick to the unintuitive (non-general) shape returned by `np.char.split`? We do have the option to "correct" it. I would say it is better to return a general shape. What does tensorflow do? In the worst-case, we could say that the op is undefined for empty inputs, but don't think that is needed here, if we can use the general shape `(2, 0, None)`.

Reply:
The key point is not about what to do for input of shape `(2, 0)` in shape-inference ... we can produce `(2, 0, 0)`. But the problem is what to do for an input of shape `(M, N)` where either `M` or `N` might or might not be zero. We would like to return `(M, N, None)` in this case.